### DISCUSSION TITLE
x-data undefined property

### DISCUSSION THREAD

**Discussion by dohspendos** [alpinejs/alpine#3301] [1. Help] [answered]
Using [alpine](https://unpkg.com/alpinejs@3.10.5).
Clicking the button results in undefined in console.

<div x-data="{
      val: 'Val Here',

      doSomething: () => {
         console.log(this.val);
      }
   }">

   <button @click='doSomething'>Do Something</button>

</div>

**Comment by dohspendos** [comment]
```html
<div x-data="{
      val: 'Val Here',

      doSomething: () => {
         console.log(this.val);
      }
   }">

   <button x-text="val" @click='doSomething'>Do Something</button>

</div>```

**Comment by firmanarz** [comment, marked as answer]
> ```
> <div x-data="{
>       val: 'Val Here',
> 
>       doSomething: () => {
>          console.log(this.val);
>       }
>    }">
> 
>    <button x-text="val" @click='doSomething'>Do Something</button>
> 
> </div>```
> ```

it's because you use arrow function on "doSomething" method, the "this" will refer to global scope, not to current object.
change your code like this, and it will work as expected.

```html
<div x-data="{
      val: 'Val Here',
      doSomething() {
         console.log(this.val);
      }
   }">
    <button x-text="val" @click='doSomething'>Do Something</button>
</div>
```

**Comment by joshhanley** [comment]
This issue has been converted to a discussion as we are now using discussions to track bugs and ideas.
See this announcement for more details https://github.com/alpinejs/alpine/discussions/1423

**Comment by ekwoka** [comment]
I'd say this is answered and done. It's how JavaScript works and I don't believe there is anything Alpine can do to solve this without forcefully transpiling the expressions.

**Reply by joshhanley** [reply]
Ok thanks, I've marked the above response as the answer
